### ISSUE TITLE
nn: the vanilla RNN layer, and whether the three recurrences want a shared base

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#185]
`LSTM` and `GRU` landed in #160. The third member of the family, the plain
Elman RNN, is missing, and ATen already has it in the shape the other two use:

```
rnn_tanh.input(Tensor input, Tensor hx, Tensor[] params, bool has_biases,
               int num_layers, float dropout, bool train, bool bidirectional,
               bool batch_first) -> (Tensor, Tensor)
rnn_relu.input(...) -> (Tensor, Tensor)
```

Two Tensor returns and a `Tensor[]` of flat weights, exactly `gru.input`'s
signature, so it needs no generator work: two allowlist lines, recipes, and
the layer. `nn.RNN` picks between them with `nonlinearity='tanh' | 'relu'`,
defaulting to tanh, so one Racket layer with `#:nonlinearity` covers both.

## Why bother, when LSTM and GRU exist

- It completes the set a reader expects from `torch.nn`, and it is the one
  every tutorial starts with.
- Its parameter naming, state shape and flat-weight order are the same as
  `GRU`'s (one gate instead of three), so it is the cheapest possible test
  that the machinery generalises rather than having been fitted to two cases.
- A single-gate recurrence is the clearest place to show the vanishing
  gradient that motivates the other two, which is worth a paragraph in the
  char-RNN example's prose.

## The base-class question

PyTorch factors the shared part into `nn.RNNBase`: parameter construction and
naming, `flatten_parameters`, `_apply`, the forward plumbing, with `RNN`,
`LSTM` and `GRU` as subclasses differing in gate count, state count and the
ATen entry point.

`torch/nn/recurrent.rkt` already has the same split, as plain values rather
than a class: an `rnn` struct holding who/gates/state-count/cudnn-mode and the
configuration, with `draw-parameters`, `ensure-flat!` and `run` taking it. Two
layers was not enough evidence to decide whether that should become something
more formal. Three is, and adding `RNN` is the moment to look at it:

- if the third layer drops in by filling out one more `rnn` spec and a
  `define-layer` form, the current factoring is right and nothing needs a base;
- if it wants to override pieces of `run` or the naming, that is the argument
  for a real shared base, and `define-layer` would need to say how one layer
  form builds on another.

Prefer the first answer. Raising it because the second would be a `define-layer`
change and therefore worth knowing about before it is written, not after.

Follows #160 (`LSTM`, `GRU`, `parameters-by-key`, rest-argument `#:forward`,
`#:on-move`) and #153 (the sequence arc). See also #184, which would remove
`#:on-move`'s only caller.
